### PR TITLE
Make multi-step PR format compulsory for bigger PRs #11

### PR DESCRIPTION
**Pull request by damithc**
Updated instructions for multi-step PRs.
Made the multi-step PRs compulsory for big PRs.

### CODE REVIEW

**Comment by damithc** [comment]
@pyokagan you are the best person to review this one. :-p Whenever you have time. No hurry.

File: docs/HowToGuides.md [hunk 1]
```diff
@@ -1,5 +1,63 @@
 # How To ...
 
+### Create a multi-step PR
+
+Big PRs (i.e. PRs touching more than 100 lines) are hard to review. 
+Using the _multi-step PR_ format explained below allows the review to be done incrementally.
+
+1. Dev creates the PR branch consisting of a series of commits corresponding to a sequence of logical steps, 
+   each step advancing the code base a step towards the end goal of the PR. The build should pass at each step.
```
Comment (pyokagan) [inline docs/HowToGuides.md]: `each step advancing the code base a step towards` -> `each step advancing the code base towards`

`The build should pass at each step.` -> `The build must pass at each step.` :-P

If the build fails at any step then it breaks bisectability.

File: docs/HowToGuides.md [hunk 2]
```diff
@@ -1,5 +1,63 @@
...
+   <details>
+   <summary>Instructions</summary>
+   
+   * There Ideal approach for creating a multi-step commit sequence:
```
Comment (pyokagan) [inline docs/HowToGuides.md]: `There Ideal` -> `The ideal`

File: docs/HowToGuides.md [hunk 3]
```diff
@@ -1,5 +1,63 @@
...
+
+1. Dev creates the PR branch consisting of a series of commits corresponding to a sequence of logical steps, 
+   each step advancing the code base a step towards the end goal of the PR. The build should pass at each step.
+   <details>
+   <summary>Instructions</summary>
+   
+   * There Ideal approach for creating a multi-step commit sequence: 
+     * Plan the PR as a sequence of logical steps.
+     * Implement each step, committing each step as a separate commit. 
+       * Write detailed commit messages as if you were explaining your commit to a future developer.
```
Comment (pyokagan) [inline docs/HowToGuides.md]: "detailed commit messages" is a different matter to "proper commit organization", I think. While it is good to have both, it is possible to have proper commit organization without detailed commit messages and vice versa.

To keep this section as succinct as possible, I think we could have a passing mention on writing good commit messages as well, but I don't think we need to dedicate a whole point and a blockquote towards it. (e.g. `Implement each step, committing each step as a separate commit. ( [Remember to write detailed commit messages!](writing-proper-commit-messages.md))`

To elaborate: "proper commit organization" mainly focuses on the *diff* of each commit. Separating logical changes into different commits allows reviewers to view the diffs of different logical changes in isolation. Detailed commit messages certainly help to *communicate* a change to the reviewer, but imo it is not strictly required for a PR to be considered a proper "multi-step PR".

File: docs/HowToGuides.md (hunk 2 above)
Comment (pyokagan) [inline docs/HowToGuides.md]: Also, this whole "ideal approach" point seems kind of redundant, given that it never happens in practice (I explain more below)

File: docs/HowToGuides.md [hunk 4]
```diff
@@ -1,5 +1,63 @@
...
+       
+        > Writing detailed commit messages is especially important for multi-step PRs because the commits will be 
+        > preserved after merging (i.e. the branch will be merged using a merge commit instead of 'squash and merge')
+        
+       * You may have to `squash` trivial commits or reorder commits to make the steps more logical. 
+   * Alternative approach:
```
Comment (pyokagan) [inline docs/HowToGuides.md]: This is actually what usually happens in practice. We may wish to note that here, in case readers try to force themselves to follow the "ideal approach".

But thinking about it, the "ideal approach" isn't really an "approach" in the first place, and we should not label it as such. Rather, it is instead the "desired end-result" of a multi-step PR.

The thing is, the thought process which we use to write code is likely not to be the same as the ordering of commits which fulfills the requirements of a multi-step PR. (e.g. bisectabilty). I like to think of it as writing a book; readers like to read a book sequentially, starting from the introduction and ending at the conclusion. However, authors are instead encouraged to write the bulk of the story first *before* writing the introduction in order to write a cohesive story.

So, overall, I think that we should not try to word this whole section as a "follow these steps 1-2-3 to get a multi-step PR!" kind of thing. Instead, I think we should

1. First explain what a good multi-step PR is, giving good examples and bad examples, to let readers understand what the desired end-result is. This is written in prose at the start of this section.

2. *Then* explain possible approaches which readers can try out to get this desired end-result, without branding any of them as "more ideal" or "more desirable", and making it clear that it is all about *personal preference*, and readers should try them out to see what they like best. This can use the 1-2-3 numbering format. From the point of view of reviewers, we don't care *how* contributors make their PR (how the sausage is made), as long as the end result is what we want.

Maybe we can also get some ideas from https://sethrobertson.github.io/GitBestPractices/#sausage

File: docs/HowToGuides.md [hunk 5]
```diff
@@ -1,5 +1,63 @@
...
+     * Implement the fix, committing after each significant change.
+     * After the fix is complete, refactor commits so the you have a commit sequence similar to what you would have 
+       achieved in the ideal approach given above.
+   </details>
+   
+1. When the PR is ready for reivew, dev posts a summary of commits using the CanIHasReview tool.
+   (created by [@pyokagan](https://github.com/pyokagan/))
```
Comment (pyokagan) [inline docs/HowToGuides.md]: As a reader, it's kind of annoying to see this actually :-P

Maybe just link `CanIHasReview` to https://github.com/pyokagan/canihasreview/ ? Enterprising readers should be able to figure out the rest.

File: docs/HowToGuides.md [hunk 6]
```diff
@@ -1,5 +1,63 @@
...
+        > preserved after merging (i.e. the branch will be merged using a merge commit instead of 'squash and merge')
+        
+       * You may have to `squash` trivial commits or reorder commits to make the steps more logical. 
+   * Alternative approach:
+     * Implement the fix, committing after each significant change.
+     * After the fix is complete, refactor commits so the you have a commit sequence similar to what you would have 
+       achieved in the ideal approach given above.
+   </details>
+   
+1. When the PR is ready for reivew, dev posts a summary of commits using the CanIHasReview tool.
```
Comment (pyokagan) [inline docs/HowToGuides.md]: `reivew` -> `review`.

File: docs/HowToGuides.md [hunk 7]
```diff
@@ -1,5 +1,63 @@
...
+       > * [v8 2/3] [create isAnyPresent() method in CollectionUtil ](https://github.com/se-edu/addressbook-level4/pull/209/commits/478ebafae4d3eb481484b0157d5f59afc1ac24ce)
+       > * [v8 3/3] [teach classes in Model to execute edit ](https://github.com/se-edu/addressbook-level4/pull/209/commits/964bc4dcf12332ce1815ca9c07bdc22f988b7825)
+       > 
+    </details>
+       
+1. The reviewer reviews one commit at a time, starting with the earliest commit. If the early commits require lot 
+   of changes, she can omit reviewing later commits until the early commits are updated as per review.
+      
+1. After reach round of review, the dev _updates the existing commits_  (rather than add more commits)
+   according to the review comments.
```
Comment (pyokagan) [inline docs/HowToGuides.md]: May wish to mention that if the dev is introducing *new* logical changes, then they should be added as separate commits rather than being squashed into existing ones.

File: docs/QuestionsIssuesPrs.md [hunk 8]
```diff
@@ -88,7 +88,9 @@ additional details provided by the specific project.
 1. After creating the PR, 
    1. Wait for CI (i.e. Travis, AppVeyor) to run tests/checks against your PR and report status. 
       If any errors are reported, fix those problems and push the fixes to the same branch.
-   1. When there are no more CI errors, post a comment to inform the dev team that the PR is ready for review. 
+   1. If the PR is big (e.g. affecting more than 100 lines), refactor the commits in the 
+      [multi-step PR format](HowToGuides.md#create-a-multi-step-pr).
+   1. When there are no more CI errors, post a comment to inform the dev team that the PR is ready for review.
```
Comment (pyokagan) [inline docs/QuestionsIssuesPrs.md]: If we are using the `CanIHasReview` tool, then there is no need to inform the dev team that the PR is ready for review. (Since the `s.ToReview` label will indicate that).

So, we may want to add a `If the multi-step PR format is not used, and there are no more CI errors, then post a comment...`

File: docs/HowToGuides.md [hunk 9]
```diff
@@ -1,5 +1,63 @@
...
+     * After the fix is complete, refactor commits so the you have a commit sequence similar to what you would have 
+       achieved in the ideal approach given above.
+   </details>
+   
+1. When the PR is ready for reivew, dev posts a summary of commits using the CanIHasReview tool.
+   (created by [@pyokagan](https://github.com/pyokagan/))  
+    <details>
+       <summary>Instructions</summary>
+       
+       * Go to `https://canihasreview.herokuapp.com/se-edu/addressbook-level4/pull/237` (change the URL to fit your PR).
```
Comment (pyokagan) [inline docs/HowToGuides.md]: There are three different components in this URL which could be different depending on repo: the org/owner (`se-edu`), the repo name (`addressbook-level4`) and the PR number (`237`). So the instruction `change the URL to fit your PR` may be vague.

Maybe we may wish to phrase it as:

1. Navigate to your PR. The URL should be something like `https://github.com/se-edu/addressbook-level4/pull/237`.

2. Replace `github.com` with `canihasreview.herokuapp.com`. The URL should be something like `https://canihasreview.herokuapp.com/se-edu/addressbook-level4/pull/237`.

(But of course the real solution is to improve the UI of CanIHasReview :-P )

**Comment by damithc** [comment]
All good comments @pyokagan Will update accordingly. 
Normal PRs (non-multi-step) are squashed before merging. Should we ask the dev to write a good description of the whole PR in the PR description (i.e. first comment) so that it can be used as the Git commit message of the squashed commit? In that case, that too should be reviewed as part of the PR review.

**Comment by pyokagan** [comment]
@damithc 
> Should we ask the dev to write a good description of the whole PR in the PR description (i.e. first comment) so that it can be used as the Git commit message of the squashed commit?

Sure. We have to remember to manually copy the PR description into the merge message box before merging, I guess.

(My personal opinion is actually to require all PRs to have proper commit organization, no matter how big or small they are. I would think that contributors who do not bother to learn how to properly use Git usually also do not submit code which meets the required quality standards anyway.)

> In that case, that too should be reviewed as part of the PR review.

Yeah.

**Comment by damithc** [comment]
@pyokagan the next version is ready for review
BTW, can you add this repo to canihasreview for me to generate commit summary?

**Comment by pyokagan** [comment]
@damithc 
> BTW, can you add this repo to canihasreview for me to generate commit summary?

Done.

**Comment by CanIHasReview** [comment]
# v1

@damithc submitted v1 for review.

* [v1 1/1] [Incorporate multi-step PRs into the process](https://github.com/oss-generic/process/pull/16/commits/0397ded17f01f15f5a6ee88dbdb0b4ab4c4e5a08)

([:books: Archive](https://github.com/oss-generic/process/compare/e252d1dafbae9fe8bc631d251bdcb7873b2a70e3...0397ded17f01f15f5a6ee88dbdb0b4ab4c4e5a08))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/oss-generic/process.git refs/pr/16/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: docs/HowToGuides.md [hunk 10]
```diff
@@ -1,5 +1,78 @@
...
+    
+    </details>
+      
+1. After reach round of review, the dev _updates the existing commits_  (rather than add more commits)
+   according to the review comments. 
+   
+   <details>
+   <summary>A typical workflow for updating the PR in response to a review</summary>
+   
+   * Commit changes in separate commits.
```
Comment (pyokagan) [inline docs/HowToGuides.md]: Maybe we should split up the points like this?

> * Updates to existing logical changes should be done by [modifying their corresponding commit](http://sethrobertson.github.io/GitPostProduction/gpp.html#analyze).
>
> * New logical changes should be introduced as new commits.
>
> * Sometimes, reviewers may recommend [splitting existing commits](http://sethrobertson.github.io/GitPostProduction/gpp.html#post-production) in order to make them more cohesive.
>
> * Commit messages should be updated with new findings from review discussions. For example, if the reviewer mentioned a possible new approach that was subsequently rejected due to a non-obvious reason, then the commit should be updated with this information.

The reason is because there are multiple ways to do them, and I think we should let this how-to be a "high level overview", For the nitty gritty of modifying a commit, we can then direct the reader to a page dedicated to that.

File: docs/HowToGuides.md [hunk 11]
```diff
@@ -1,5 +1,78 @@
...
+### Create a multi-step PR
+
+> Ask a programmer to review 10 lines of code, he'll find 10 issues. Ask him to do 500 lines and he'll say it 
+> looks good. _[[source](https://twitter.com/girayozil/status/306836785739210752)]_
+
+Big PRs (i.e. PRs touching more than 100 lines) are hard to review. Using the _multi-step PR_ format allows the 
+review to be done incrementally.
+
+A multi-step PR consists a series of commits corresponding to a sequence of logical steps, 
+each step advancing the code base towards the end goal of the PR. The build must pass at each step.
```
Comment (pyokagan) [inline docs/HowToGuides.md]: Some suggested additions to the criteria of what makes a good multi-step PR. (Need to tweak the wording to make them read better as required)

* Each commit contains a single logical change, and this change must stand on its own. (or, you can word it as: each commit has a single responsibility, and this responsibility must be fully carried out) (i.e. in AddressBook, a commit that claims it is implementing an `EditCommand` cannot have its `EditCommand::execute()` method throw a "Not Implemented" error, since this means that the commit did not *fully* implement the `EditCommand` which is at odds with what it claims to do. We can't use this example in a document that's supposed to be generic though :-P)

* The series of commits in the PR are ordered in a bottom-up fashion, each commit building on top of each other towards the end goal of the PR.

* The build passes at each commit of the PR.

* Each commit has a detailed commit message which explains the context and rationale behind the commit. See the writing commit messages doc for more details.

Comment (pyokagan) [inline docs/HowToGuides.md]: Also, we may wish to show some hypothetical examples on what is not a good multi-step PR, in order to aid the reader's understanding. But let's leave it for later since that takes some effort to do.

**Comment by damithc** [comment]
@pyokagan I'll merge now (on account of code sprint) and improve as suggested in a separate PR
